FfiSpike/FfiSpike.cpp

```diff
@@ -877,52 +895,93 @@ bool ensure_device_path_exists(std::string &device_path, afc_connection *connect
 	return true;
 }
 
-void upload_file(std::string device_identifier, const char *application_identifier, const char *source, const char *destination, std::string method_id) {
-	afc_file_ref file_ref;
-	std::string destination_str =  windows_path_to_unix(destination);
-	afc_connection* afc_conn_p = get_afc_connection(device_identifier, application_identifier, destination_str, method_id);
+void accumulate_errors(const std::function<unsigned ()>& expression_to_evaluate, const std::string& error_message, std::vector<std::string>& errors_container)
+{
+	if (expression_to_evaluate())
+	{
+		errors_container.push_back(error_message);
+	}
+}
+
+void upload_file(std::string device_identifier, const char *application_identifier, const std::vector<FileUploadData>& files, std::string method_id) {
+	std::string afc_destination_str = windows_path_to_unix(files[0].destination);
+	afc_connection* afc_conn_p = get_afc_connection(device_identifier, application_identifier, afc_destination_str, method_id);
 	if (!afc_conn_p)
 	{
+		// If there is no opened afc connection the get_afc_connection will print the error for the operation.
 		return;
 	}
 
-	std::ifstream file(source, std::ios::binary | std::ios::ate);
-	std::streamsize size = file.tellg();
-	file.seekg(0, std::ios::beg);
-
-	std::vector<char> file_content(size);
-	if (file.read(file_content.data(), size))
+	std::vector<std::string> errors;
+	std::vector<std::thread> file_upload_threads;
+	
+	for (size_t i = 0; i < files.size(); i++)
 	{
-		destination = destination_str.c_str();
-		std::string dir_name = get_dirname(destination);
-		if (ensure_device_path_exists(dir_name, afc_conn_p, method_id, device_identifier))
+		file_upload_threads.emplace_back([=, &errors]() -> void
 		{
-			std::stringstream error_message;
-			AFCRemovePath(afc_conn_p, destination);
-			error_message << "Could not open file " << destination << " for writing";
-			PRINT_ERROR_AND_RETURN_IF_FAILED_RESULT(AFCFileRefOpen(afc_conn_p, destination, kAFCFileModeWrite, &file_ref), error_message.str().c_str(), device_identifier, method_id);
-			error_message.str("");
-			error_message << "Could not write to file: " << destination;
-			PRINT_ERROR_AND_RETURN_IF_FAILED_RESULT(AFCFileRefWrite(afc_conn_p, file_ref, &file_content[0], file_content.size()), error_message.str().c_str(), device_identifier, method_id);
-			error_message.str("");
-			error_message << "Could not close file reference: " << destination;
-			PRINT_ERROR_AND_RETURN_IF_FAILED_RESULT(AFCFileRefClose(afc_conn_p, file_ref), error_message.str().c_str(), device_identifier, method_id);
-			error_message.str("");
-			print(json({{ kResponse, "Successfully uploaded file" },{ kId, method_id },{ kDeviceId, device_identifier }}));
-		}
+			FileUploadData current_file_data = files[i];
+			afc_file_ref file_ref;
+			std::string source = current_file_data.source;
+			std::string destination = current_file_data.destination;
+
+			std::ifstream file(source, std::ios::binary | std::ios::ate);
```

Comment:
You can use `get_file_info` from `FileHelper.h` here in order to get the file size along with its contents